### PR TITLE
Adding proxy support

### PR DESCRIPTION
**Pull request by phawxby**
Without a disgusting amount of dependencies

Fix for bugs #448 & #337

### CODE REVIEW

File: utils/ChromiumDownloader.js [hunk 1]
```diff
@@ -216,3 +221,16 @@ function downloadFile(url, destinationPath, progressCallback) {
 function extractZip(zipPath, folderPath) {
   return new Promise(fulfill => extract(zipPath, {dir: folderPath}, fulfill));
 }
+
+function applyAgent(opt)
+{
+  if (process.env.http_proxy || process.env.https_proxy)
```
Comment (aslushnikov) [inline utils/ChromiumDownloader.js]: nit: put `{` on the same line

File: utils/ChromiumDownloader.js [hunk 2]
```diff
@@ -216,3 +221,16 @@ function downloadFile(url, destinationPath, progressCallback) {
 function extractZip(zipPath, folderPath) {
   return new Promise(fulfill => extract(zipPath, {dir: folderPath}, fulfill));
 }
+
+function applyAgent(opt)
```
Comment (aslushnikov) [inline utils/ChromiumDownloader.js]: nit: put `{` on the same line

```js
function applyAgent(opt) {
...
```

File: utils/ChromiumDownloader.js [hunk 3]
```diff
@@ -216,3 +221,16 @@ function downloadFile(url, destinationPath, progressCallback) {
...
+{
+  if (process.env.http_proxy || process.env.https_proxy)
+  {
+    let parsedProxy = process.env.http_proxy ? URL.parse(process.env.http_proxy) : URL.parse(process.env.https_proxy);
```
Comment (aslushnikov) [inline utils/ChromiumDownloader.js]: A few questions:
1. we should support capitalized env variables as well (and camel-cased, probably). There's also the NO_PROXY env variable, not sure how often it is used. Maybe rely on https://github.com/Rob--W/proxy-from-env to parse environment? 
2. we only do a single request to https url - do we care about `http_proxy` at all?

Comment (tardyp) [reply utils/ChromiumDownloader.js]: 👍  please do not reinvent another way of parsing the proxy env variables.

**Comment by phawxby** [comment]
Thanks for the feedback @aslushnikov. How about this?

File: utils/ChromiumDownloader.js [hunk 4]
```diff
@@ -217,3 +224,17 @@ function downloadFile(url, destinationPath, progressCallback) {
 function extractZip(zipPath, folderPath) {
   return new Promise(fulfill => extract(zipPath, {dir: folderPath}, fulfill));
 }
+
+function applyAgent(opt) {
+  const proxy_url = getProxyForUrl(opt.href);
+
+  if (proxy_url)
+  {
```
Comment (aslushnikov) [inline utils/ChromiumDownloader.js]: let's move this to the same line as `if` and call variable `proxyURL`

File: utils/ChromiumDownloader.js [hunk 5]
```diff
@@ -217,3 +224,17 @@ function downloadFile(url, destinationPath, progressCallback) {
 function extractZip(zipPath, folderPath) {
   return new Promise(fulfill => extract(zipPath, {dir: folderPath}, fulfill));
 }
+
+function applyAgent(opt) {
```
Comment (aslushnikov) [inline utils/ChromiumDownloader.js]: nit: let's call parameter `options`, we tend to avoid abbreviations

File: utils/ChromiumDownloader.js [hunk 6]
```diff
@@ -217,3 +224,17 @@ function downloadFile(url, destinationPath, progressCallback) {
...
+  const proxy_url = getProxyForUrl(opt.href);
+
+  if (proxy_url)
+  {
+    const parsedProxy = URL.parse(proxy_url);
+
+    if (parsedProxy.protocol === 'http:')
+      parsedProxy.secureProxy = false;
```
Comment (aslushnikov) [inline utils/ChromiumDownloader.js]: AFAIU secureProxy is never true. You rather want:

```js
parsedProxy.secureProxy = parsedProxy.protocol === 'https:';
```

File: utils/ChromiumDownloader.js [hunk 7]
```diff
@@ -217,3 +224,17 @@ function downloadFile(url, destinationPath, progressCallback) {
...
+
+    opt.agent = new ProxyAgent(parsedProxy);
```
Comment (aslushnikov) [inline utils/ChromiumDownloader.js]: I don't like that we pass result of URL.parse over to the ProxyAgent constructor. I had to check what's the output of `URL.parse` and make sure it doesn't clash with some options of `ProxyAgent`.

Let's make a method `requestOptions` which returns options for the `https.request` module, and will handle proxy inside. Something along the lines: 

```js
function requestOptions(url, method = 'GET') {
  const parsedURL = URL.parse(url);
  const result = {
    method,
    host: parsedURL.host,
    path: parsedURL.pathname
  };
  const proxyURL = getProxyForUrl(url);
  if (proxyURL) {
    const parsedProxyURL = URL.parse(proxyURL);
    result.agent = new ProxyAgent({
      host: parsedProxyURL.host,
      port: parsedProxyURL.port,
      secureProxy: parsedProxyURL.protocol === 'https:'
    });
  }
  return result;
}
```

**Review by aslushnikov — COMMENTED** [review]
This is great! A few comments before we can have this landed.

File: utils/ChromiumDownloader.js (hunk 7 above)
Comment (phawxby) [reply utils/ChromiumDownloader.js]: Constructing a result object seems unnecessary, https.get and https.request accept a URL object and [internally it makes use of urlToOptions](https://github.com/nodejs/node/blob/master/lib/https.js#L234), it makes sense to [leave node to handle the conversion](https://github.com/nodejs/node/blob/master/lib/internal/url.js#L1298) to a options object should anything change in future. 

The same applies to `ProxyAgent`, [it uses url.parse internally too](https://github.com/TooTallNate/node-https-proxy-agent/blob/master/index.js#L27). When I switched it to using the constructed object as you suggested above it broke.

**Comment by phawxby** [comment]
Once Travis does his thing hopefully that'll be it. Side note, linting absolutely explodes when run on windows to the point of being useless which is why I need to check in first. It would be handy if linting worked on Windows.

**Review by aslushnikov — APPROVED** [review]
Thanks a lot for the patch!

**Comment by phawxby** [comment]
@aslushnikov do you guys have a release schedule? I'm just wondering when this will be deployed so I can start using it.

**Comment by aslushnikov** [comment]
@phawxby we are still figuring out the process, so no schedule yet. But the next release should happen shortly

**Pull request merged**